algorithms/src/snark/marlin/ahp/prover/round_functions/fourth.rs

```diff
@@ -12,58 +12,229 @@
+        for (&circuit, state_i) in &state.circuit_specific_states {
+            let v_R_i_at_alpha = state_i.constraint_domain.evaluate_vanishing_polynomial(*alpha);
+            let v_C_i_at_beta = state_i.variable_domain.evaluate_vanishing_polynomial(*beta);
+            let v_R_i_alpha_v_C_i_beta = v_R_i_at_alpha * v_C_i_at_beta;
+            let k_domains = [state_i.non_zero_a_domain, state_i.non_zero_b_domain, state_i.non_zero_c_domain];
+            let ariths = [&circuit.a_arith, &circuit.b_arith, &circuit.c_arith];
+
+            for (matrix_label, non_zero_domain, arith) in itertools::izip!(matrix_labels, k_domains, ariths) {
+                let label = witness_label(circuit.id, matrix_label, 0);
+                pool.add_job(move || {
+                    let result = Self::calc_matrix_sumcheck_witness(
+                        label,
+                        state_i.constraint_domain,
+                        state_i.variable_domain,
+                        non_zero_domain,
+                        arith,
+                        *alpha,
+                        *beta,
+                        v_R_i_alpha_v_C_i_beta,
+                        max_non_zero_domain_size,
+                        &circuit.fft_precomputation,
+                        &circuit.ifft_precomputation,
+                    );
+                    (circuit, result)
+                });
+            }
+        }
+
+        let mut sums = Vec::with_capacity(state.circuit_specific_states.len());
+        let mut gs = BTreeMap::new();
+        for ((circuit_a, results_a), (circuit_b, results_b), (circuit_c, results_c)) in
+            pool.execute_all().into_iter().tuples()
+        {
+            assert_eq!(circuit_a, circuit_b);
+            assert_eq!(circuit_a, circuit_c);
+            let (sum_a, lhs_a, g_a) = results_a?;
+            let (sum_b, lhs_b, g_b) = results_b?;
+            let (sum_c, lhs_c, g_c) = results_c?;
+            let matrix_sum = prover::message::MatrixSums { sum_a, sum_b, sum_c };
+            sums.push(matrix_sum);
+            state.circuit_specific_states.get_mut(circuit_a).unwrap().lhs_polynomials = Some([lhs_a, lhs_b, lhs_c]);
+            let matrix_gs = prover::MatrixGs { g_a, g_b, g_c };
+            gs.insert(circuit_a.id, matrix_gs);
+        }
+
+        let msg = prover::FourthMessage { sums };
+        let oracles = prover::FourthOracles { gs };
+
+        assert!(oracles.matches_info(&Self::fourth_round_polynomial_info(
+            state.circuit_specific_states.keys().map(|c| (c.id, &c.index_info))
+        )));
+
+        end_timer!(round_time);
+
+        Ok((msg, oracles, state))
     }
 
-    /// Output the degree bounds of oracles in the third round.
-    pub fn fourth_round_polynomial_info() -> BTreeMap<PolynomialLabel, PolynomialInfo> {
-        [("h_2".into(), PolynomialInfo::new("h_2".into(), None, None))].into()
+    #[allow(clippy::too_many_arguments)]
+    fn calc_matrix_sumcheck_witness(
```

Comment:
Consistent with `calculate_mask_poly`.

```suggestion
    fn calculate_matrix_sumcheck_witness(
```

Reply:
955fe6d13